handler.go

```diff
@@ -0,0 +1,276 @@
+	}
+	// Second and third parameters are common to all three forms.
+	if t.In(1) != paramsType {
+		return nil, errgo.Newf("second argument is %s, need %s", t.In(1), paramsType)
+	}
+	pt, err := getRequestType(t.In(2))
+	if err != nil {
+		return nil, errgo.Notef(err, "third argument cannot be used for Unmarshal")
+	}
+	if t.NumOut() < 2 {
+		if t.In(0) != httpResponseWriterType {
+			return nil, errgo.Newf("first argument is %s, need http.ResponseWriter", t.In(0))
+		}
+	} else {
+		if t.In(0) != httpHeaderType {
+			return nil, errgo.Newf("first argument is %s, need http.Header", t.In(0))
+		}
+	}
+	if t.NumOut() > 0 {
+		if et := t.Out(t.NumOut() - 1); et != errorType {
+			return nil, errgo.Newf("final result parameter is %s, need error", et)
+		}
+	}
+	return pt, nil
+}
+
+func handleParams(fv reflect.Value, pt *requestType) func(w http.ResponseWriter, req *http.Request, p httprouter.Params) ([]reflect.Value, error) {
+	ft := fv.Type()
+	arg0IsHeader := ft.NumOut() == 2
+	argStructType := ft.In(2).Elem()
+	return func(w http.ResponseWriter, req *http.Request, p httprouter.Params) ([]reflect.Value, error) {
+		if err := req.ParseForm(); err != nil {
+			return nil, errgo.WithCausef(err, ErrUnmarshal, "cannot parse HTTP request form")
+		}
+		params := Params{
+			Request: req,
+			PathVar: p,
+		}
+		pv := reflect.New(argStructType)
+		if err := unmarshal(params, pv, pt); err != nil {
+			return nil, errgo.NoteMask(err, "cannot unmarshal parameters", errgo.Is(ErrUnmarshal))
+		}
+		var arg0 reflect.Value
+		if arg0IsHeader {
+			arg0 = reflect.ValueOf(w.Header())
+		} else {
+			arg0 = reflect.ValueOf(w)
+		}
+		return fv.Call([]reflect.Value{
+			arg0,
+			reflect.ValueOf(params),
+			pv,
+		}), nil
+	}
+}
+
+func (e ErrorMapper) handleResult(ft reflect.Type, f func(w http.ResponseWriter, req *http.Request, p httprouter.Params) ([]reflect.Value, error)) httprouter.Handle {
+	switch ft.NumOut() {
+	case 0:
+		return func(w http.ResponseWriter, req *http.Request, p httprouter.Params) {
+			_, err := f(w, req, p)
+			if err != nil {
+				e.WriteError(w, err)
+			}
+		}
+	case 1:
+		return func(w http.ResponseWriter, req *http.Request, p httprouter.Params) {
+			out, err := f(w, req, p)
+			if err != nil {
+				e.WriteError(w, err)
+				return
+			}
+			herr := out[0].Interface()
+			if herr != nil {
+				e.WriteError(w, herr.(error))
+			}
+		}
+	case 2:
+		return func(w http.ResponseWriter, req *http.Request, p httprouter.Params) {
+			out, err := f(w, req, p)
+			if err != nil {
+				e.WriteError(w, err)
+				return
+			}
+			herr := out[1].Interface()
+			if herr != nil {
+				e.WriteError(w, herr.(error))
+				return
+			}
+			err = WriteJSON(w, http.StatusOK, out[0].Interface())
+			if err != nil {
+				e.WriteError(w, err)
+			}
+		}
+	default:
+		panic("unreachable")
+	}
+}
+
+// ToHTTP converts an httprouter.Handle into an http.Handler.
+// It will pass no path variables to h.
+func ToHTTP(h httprouter.Handle) http.Handler {
```

Comment:
What if there is just a function that accepts interface and returns http.Handler instead of having two func calls?